### ISSUE TITLE
About Improved SemHash

### ISSUE THREAD

**Issue by ycjing** [HB-Liu/Improved-SemHash-and-Gumbel-Softmax#1]
Hi @R00Kie-Liu 

Thanks for this great code! When I run the code of improved semhash, I found that the output would sometimes become continuous values, instead of discrete decision. I'm not so familiar with improved semhas. May I ask if this is normal? Thanks so much!

Cheers,

**Comment by R00Kie-Liu** [comment]
Yes, its normal. 
For Improved semantic hashing method:
During training, in the forward pass the authors **use v1 half of the time and v2 the other half**. 
In the backward pass, the gradients always flow to v1, even if we used v2 in the forward computation.

More details can be refer to the orginal paper.

**Comment by ycjing** [comment]
Hi @R00Kie-Liu 

Thank you so much for the quick response! I deeply appreciate it. Also, thanks for the great implementation again!

I have one more minor question. For Gumbel Softmax, the output results are all one-hot vectors. However, from the results of improved SemHash, it seems that the outputs are not always one-hot. Specifically, there may be many 1s and it is also possible that there are no 1s. May I ask if this behavior is normal? Thanks so much! Have a nice day!

Cheers,
Yongcheng

**Comment by R00Kie-Liu** [comment]
It's normal, I think. In this way, these two methods can be applied to different scenes. 

For Gumbel max, since it outputs a one-hot vector, it can be used to implement the hard selection operation during training. Also, the Gumbel max trick can be easily extended to multiple 1s outputs (switch to many binary vectors).

**Comment by ycjing** [comment]
Hi @R00Kie-Liu 

Thank you for the warm help! I appreciate it. Now I understand the whole thing. Thanks!
